### PR TITLE
CXX-2022 implement encryption corpus tests

### PR DESCRIPTION
**Pull request by samantharitter**

### CODE REVIEW

File: src/mongocxx/test/client_side_encryption.cpp [hunk 1]
```diff
@@ -681,8 +689,292 @@ TEST_CASE("Views are prohibited", "[client_side_encryption]") {
...
+            // If allowed is false, validate the value exactly equals the value of the corresponding
+            // field of corpus (neither was encrypted).
+            REQUIRE(value == corpus.view()[field_name]["value"].get_value());
+        }
+    }
+}
+
 TEST_CASE("Corpus", "[client_side_encryption]") {
-    // TODO CXX-2022
+    if (!mongocxx::test_util::should_run_client_side_encryption_test()) {
```
Comment (samantharitter) [inline src/mongocxx/test/client_side_encryption.cpp:972]: Looks like this needs the server version skip, too, to run successfully against 4.0 servers on evergreen

File: src/mongocxx/test/client_side_encryption.cpp [hunk 2]
```diff
@@ -49,6 +51,9 @@ const auto kLocalMasterKey =
     "\x4f\x69\x37\x36\x36\x4a\x7a\x58\x5a\x42\x64\x42\x64\x62\x64\x4d\x75\x72"
     "\x64\x6f\x6e\x4a\x31\x64";
 
+const auto kLocalKeyUUID = "\x2c\xe0\x80\x2c\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00";
```
Comment (kevinAlbs) [inline src/mongocxx/test/client_side_encryption.cpp:54]: Can you comment noting this is a base64 decoding of `LOCALAAAAAAAAAAAAAAAAA==` and the value below a decoding of `AWSAAAAAAAAAAAAAAAAAAA==`? That will be easier to match up with the spec.

File: src/mongocxx/test/client_side_encryption.cpp [hunk 3]
```diff
@@ -681,8 +689,306 @@ TEST_CASE("Views are prohibited", "[client_side_encryption]") {
...
+
+    auto db = client["db"];
+    auto cmd = document{} << "create"
+                          << "coll"
+                          << "validator" << open_document << "$jsonSchema" << corpus_schema.view()
+                          << close_document << finalize;
+
+    db.run_command(cmd.view());
+
+    // Using client, drop the collection keyvault.datakeys. Insert the documents
```
[the remark this answers is not in the document]
Comment (samantharitter) [reply src/mongocxx/test/client_side_encryption.cpp:730]: Done.

File: src/mongocxx/test/client_side_encryption.cpp [hunk 4]
```diff
@@ -681,8 +689,306 @@ TEST_CASE("Views are prohibited", "[client_side_encryption]") {
...
+    memcpy(&(local_master_key_storage[0]), kLocalMasterKey, 96);
+
+    bsoncxx::types::b_binary local_key_id{
+        bsoncxx::binary_sub_type::k_uuid, 16, (const uint8_t*)&local_key_id_storage};
+    bsoncxx::types::b_binary aws_key_id{
+        bsoncxx::binary_sub_type::k_uuid, 16, (const uint8_t*)&aws_key_id_storage};
+    bsoncxx::types::b_binary local_master_key{
+        bsoncxx::binary_sub_type::k_binary, 96, (const uint8_t*)&local_master_key_storage};
+
+    auto local_key_value = make_value(local_key_id);
```
[the remark this answers is not in the document]
Comment (samantharitter) [reply src/mongocxx/test/client_side_encryption.cpp:761]: ty!

File: src/mongocxx/test/client_side_encryption.cpp (hunk 2 above)
Comment (samantharitter) [reply src/mongocxx/test/client_side_encryption.cpp:54]: Done.

File: src/mongocxx/test/client_side_encryption.cpp [hunk 5]
```diff
@@ -681,8 +689,292 @@ TEST_CASE("Views are prohibited", "[client_side_encryption]") {
...
+    auto corpus = _doc_from_file("/corpus/corpus.json");
+    auto corpus_encrypted_expected = _doc_from_file("/corpus/corpus-encrypted.json");
+
+    // Create a new BSON document, named corpus_copied.
+    auto corpus_copied_builder = bsoncxx::builder::basic::document{};
+
+    // Iterate over each field of corpus.
+    for (bsoncxx::document::element ele : corpus.view()) {
+        auto field_name_view = ele.key();
+        std::string field_name{field_name_view};
```
[the remark this answers is not in the document]
Comment (samantharitter) [reply src/mongocxx/test/client_side_encryption.cpp:803]: That's a great idea, nice

**Pull request merged**